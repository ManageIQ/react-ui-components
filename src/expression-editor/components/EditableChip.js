import React from 'react';
import PropTypes from 'prop-types';
import Chip from './Chip';

import AutocompleteTextInput from './AutocompleteTextInput';
import ExpressionEditorPropTypes from './ExpressionEditorPropTypes';

class EditableChip extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      filteredOptions: props.options,
      filterString: this.props.label,
      aliasMode: false,
      // filterString: '',
    };
    // console.log('editable chip constructor', props);
  }

  static getDerivedStateFromProps(props, state) {
    // console.log('editable chipget derived state', props, state);
    return { filteredOptions: props.options.filter(props.filterOption(state.filterString)) };
  }

  onClick = () => {
    this.props.onClick(this.props.index);
  };

  onDoubleClick = () => {
    this.setState({ filterString: this.props.label });
    this.props.onDoubleClick(this.props.selected, this.props.index);
  };

  onSubmit = (selected) => {
    // console.log('EditableChip submit', selected, this.props.item);
    console.log('MODE:', this.state.aliasMode);
    if (this.state.aliasMode) {
      this.setAlias(selected);
      this.setState({aliasMode: false});
    } else {
      const newSelected = selected;
      if (selected.type === 'userinput') {
        // console.log('SUMBIT:',selected);
        newSelected.parent = this.props.item.parent;
      }
      // let{ label, id, item } = this.props;
      this.setState({ filterString: '' });
      this.props.onSubmit(newSelected, this.props.item);
    }
  }

  onDelete = () => {
    // console.log('on DELETEE');
    this.props.onDelete(this.props.item, this.props.index);
  }

  onFocus = () => {
    // console.log('on FOCUSSSSSSSS');
    this.props.onFocus(this.props.item, this.props.index);
  }

  onBlur = () => {
    // console.log('on bluuuuuuuuuuuuuuuuuur');
    this.props.onBlur(this.props.item, this.props.index);
  }

  onUserInput = (input) => {
  // console.log('EditableChip UserInput', input);
    this.setState({ filterString: input });
  }

  onKeyDown = (key) => {
    this.props.onKeyDown(key, this.props.index, this.props.selected);
  }

  setAliasMode = (bool) => {
    this.setState({aliasMode: bool});
  }

  setAlias = (alias) => {
    console.log('ALIAS',alias, this.props.index );
    this.props.setAlias(alias, this.props.index)
  }

  render() {
    // console.log('EditableChip props:', this.props.label);
    return (
      (this.props.isEditing &&
        <AutocompleteTextInput
          onSubmit={this.onSubmit}
          onChange={this.onUserInput}
          onKeyDown={this.onKeyDown}
          options={this.state.filteredOptions}
          value={this.state.filterString}
          denyUserInput={(this.props.options.length > 0)}
          // setAlias={this.props.setAlias}
          setAliasMode={this.setAliasMode}
          aliasMode={this.state.aliasMode}
          isLastElement={!!this.props.isLastElement}
          deleteExpression={this.props.deleteExpression}
          blurAllChips={this.props.blurAllChips}
          // registerInput={this.props.registerInput}
          // unregisterInput={this.props.unregisterInput}
          inputRef={this.props.inputRef}
          index={this.props.index}
          next={this.props.item.next}
        />) ||
        <Chip
          onDelete={this.onDelete}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onKeyDown={this.onKeyDown}
          registerChip={this.props.registerChip}
          unregisterChip={this.props.unregisterChip}
          isAliasSet={this.props.isAliasSet}
          setAliasMode={this.setAliasMode}
          setAlias={this.setAlias}
          isFocused={this.props.isFocused}
          index={this.props.index}
          onClick={this.onClick}
          onDoubleClick={this.onDoubleClick}
          label={this.props.label}
          type={this.props.selected.type}
          chipRef={this.props.chipRef}
        />
    );
  }
}

EditableChip.propTypes = {
  onClick: PropTypes.func.isRequired,
  onDoubleClick: PropTypes.func.isRequired,
  onSubmit: PropTypes.func,
  onKeyDown: PropTypes.func,
  onDelete: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  setAlias: PropTypes.func,
  deleteExpression: PropTypes.func,
  registerChip: PropTypes.func,
  unregisterChip: PropTypes.func,
  label: PropTypes.string.isRequired,
  item: PropTypes.object,
  index: PropTypes.number.isRequired,
  isAliasSet: PropTypes.bool,
  isFocused: PropTypes.bool.isRequired,
  isEditing: PropTypes.bool,
  options: PropTypes.arrayOf(ExpressionEditorPropTypes.term).isRequired,
  selected: ExpressionEditorPropTypes.option,
  filterOption: PropTypes.func,
  inputRef: PropTypes.object,
  chipRef: PropTypes.object,
};

EditableChip.defaultProps = {
  item: {},
  selected: {},
  isEditing: true,
  filterOption: value => option => option.label.toLowerCase().includes(value.toLowerCase()),
};


export default EditableChip;
