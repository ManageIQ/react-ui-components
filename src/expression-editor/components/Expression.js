import React from 'react';
import PropTypes from 'prop-types';
import EditableChip from './EditableChip';

import ExpressionEditorPropTypes from './ExpressionEditorPropTypes';

class Expression extends React.Component {
  onClick = (chipIndex) => {
    console.log('expression onclick');
    this.props.onClick(this.props.index, chipIndex);
    // this.setState({isEditing: !selected.isEditing});
    // console.log(selected);
  }

  onDoubleClick = (selected, chipIndex) => {
    // console.log('EditableChip onDoubleClickclick');
    this.props.onDoubleClick(selected, this.props.index, chipIndex);
  }

  onSubmit = (selected, previous) => {
    // console.log('expression on submit', selected, previous, this.props);
    this.props.onSubmit(selected, previous, this.props.index);
  }

  onDelete = (selected, chipIndex) => {
    // console.log('ON DELETE', selected);
    this.props.onDelete(selected, this.props.index, chipIndex);
  }

  onFocus = (selected, chipIndex) => {
    // console.log('ON FOCUS', selected);
    this.props.onFocus(selected, this.props.index, chipIndex);
  }

  onBlur = (selected, chipIndex) => {
    // console.log('ON BLUR', selected);
    this.props.onBlur(selected, this.props.index, chipIndex);
  }

  onKeyDown = (key, index, selected) => {
    // console.log('EXRESSION ON KEY DOWN', index);
    this.props.onKeyDown(key, index, selected, this.props.index);
  }

  onInsertExpressionClick = () => {
    this.props.onInsertExpression(this.props.index);
  }

  onDeleteExpressionClick = () => {
    this.props.onDeleteExpression(this.props.index);
  }


  generateChip = ({ term, flags }, index) => {
    // console.log('AAAAAAAAA', term, flags, index, this.props.chipRefs);
    return (<li className="chip">
      <EditableChip
        key={term.id}
        label={flags.alias || term.label}
        isEditing={flags.isEditing}
        isFocused={flags.isFocused}
        onClick={this.onClick}
        onDoubleClick={this.onDoubleClick}
        onSubmit={this.onSubmit}
        onDelete={this.onDelete}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
        setAlias={this.setAlias}
        blurAllChips={this.props.blurAllChips}
        // onKeyDown={this.props.onKeyDown}
        chipRef={this.props.chipRefs[index]}
      // registerInput={this.props.registerInput}
      // unregisterInput={this.props.unregisterInput}
        options={term.parent.next}
        selected={term}
        item={term}
        index={index}
        onKeyDown={this.onKeyDown}
      />
    </li>)
  }


  render() {
    // console.log('Expression options', this.props.next.parent.next);
    // const { this: { props: { expression } } } = this.props.expression;
    // const {this: {props: {expression: expression}}} = this;
    const expression = this.props.expression;
    const options = this.props.parenthesesCount.left > this.props.parenthesesCount.right ? this.props.next.parent.next : this.props.next.parent.next.filter(t => t.type !== "rightParenteze");
    // console.log('expression props', expression);
    const newChip = (
      <li>
        <EditableChip
          label={this.props.next.label}
          isEditing={this.props.next.isEditing}
          onClick={this.onClick}
          onDoubleClick={this.onDoubleClick}
          onSubmit={this.onSubmit}
          onFocus={this.onFocus}
          onBlur={this.onBlur}
          onKeyDown={this.onKeyDown}
          deleteExpression={this.onDeleteExpressionClick}
          blurAllChips={this.props.blurAllChips}
         // registerInput={this.props.registerInput}
         // unregisterInput={this.props.unregisterInput}
          inputRef={(this.props.isLastExpression ? this.props.inputRef : {})}
          options={options}
          selected={this.props.next}
          item={this.props.next}
          index={this.props.expression.length}
          isLastElement={this.props.isLastExpression}
        />
      </li>);
    const isLastEditing = (expression[expression.length - 1] && expression[expression.length - 1].flags.isEditing);
    const endOfExpresion = this.props.next.parent.next.length === 0;
    // const isFocused = expression.map(t => t.flags.isFocused).reduce((a, b) => (a || b), false);
    // console.log('IS FOCUSED', this.props.isFocused);
    // console.log('Expression',endOfExpresion, this.props.next.parent);
    // console.log('expression props next', this.props.next, isLastEditing);

    return (
      <React.Fragment>
        <ul className="list-inline expression">
          {expression.map((term, index) => this.generateChip(term, index))}
          {/* {(isLastEditing || endOfExpresion || null) || newChip} */}
          {(endOfExpresion || isLastEditing || null) || ((this.props.isFocused || this.props.isLastExpression) && newChip)}
          {/* {newChip} */}
          <span className="buttons">
            <button onClick={this.onDeleteExpressionClick}> Delete Expression </button>
            <button onClick={this.onInsertExpressionClick}> Add Expression </button>
          </span>
        </ul>
      </React.Fragment>
    );
  }
}


Expression.propTypes = {
  // isEditing: PropTypes.arrayOf(),
  onClick: PropTypes.func.isRequired,
  onDoubleClick: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onKeyDown: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onFocus: PropTypes.func.isRequired,
  onBlur: PropTypes.func.isRequired,
  onInsertExpression: PropTypes.func.isRequired,
  onDeleteExpression: PropTypes.func.isRequired,
  expression: ExpressionEditorPropTypes.expression,
  next: ExpressionEditorPropTypes.term,
  index: PropTypes.number.isRequired,
  isFocused: PropTypes.bool.isRequired,
  chipRefs: PropTypes.arrayOf(PropTypes.object),
  inputRef: PropTypes.object,
};


export default Expression;
