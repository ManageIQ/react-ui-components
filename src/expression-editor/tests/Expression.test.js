import React from 'react';
import Expression from '../components/Expression';
import EditableChip from '../components/EditableChip';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const expressionProps = {
  key: 0,
  onClick: jest.fn(),
  onDoubleClick: jest.fn(),
  onSubmit: jest.fn(),
  onKeyDown: jest.fn(),
  onDelete: jest.fn(),
  onFocus: jest.fn(),
  onBlur: jest.fn(),
  blurAllChips: jest.fn(),
  setAlias: jest.fn(),
  onInsertExpression: jest.fn(),
  onDeleteExpression: jest.fn(),
  expression: [{flags: {isEditing: false}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}],
  chipRefs: [React.createRef()],
  isFocused: false,
  index: 1,
  isLastExpression: false,
  parenthesesCount: {left:0, right: 0},
  inputRef: React.createRef(),
  next: {id: 2, label: 'input', type: 'userinput', next:[{id: 2, label: 'input', type: 'userinput', next:[], parent: null}], parent: {id: 3, label: 'ahoj', type: 'userinput', next:[], parent: null}}
};

describe('Expression Component', () => {

  it('Onclick calls callback', () => {
    let onClick = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps} onClick={onClick}/>
    );

    wrapper.find('span.pf-c-label').simulate('click');
    expect(onClick).toBeCalledWith(1, 0);
  });

  it('onDoubleClick calls callback', () => {
    let onDoubleClick = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps} onDoubleClick={onDoubleClick}/>
    );

    wrapper.find('span.pf-c-label').simulate('dblclick');
    expect(onDoubleClick).toBeCalledWith({"id": 1, "label": "ahoj", "next": [], "parent": {"id": 0, "label": "root", "next": [], "parent": null, "type": "root"}, "type": "category"}, 1, 0);
  });

  it('onSubmit calls callback', () => {
    let onSubmit = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: true}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        isFocused
        onSubmit={onSubmit}
      />
    );

    wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
    expect(onSubmit).toBeCalledWith(
      {"id": "ahoj", "label": "ahoj", "next": [], "parent": {"id": 0, "label": "root", "next": [], "parent": null, "type": "root"}, "type": "userinput"},
      {"id": 1, "label": "ahoj", "next": [], "parent": {"id": 0, "label": "root", "next": [], "parent": null, "type": "root"}, "type": "category"},
      1
    );
  });

  it('onDelete calls callback', () => {
    let onDelete = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: false, isFocused: true}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        onDelete={onDelete}
        isFocused
      />
    );

    wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
    wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
    wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
    expect(onDelete).toBeCalledWith( {"id": 1, "label": "ahoj", "next": [], "parent": {"id": 0, "label": "root", "next": [], "parent": null, "type": "root"}, "type": "category"}, 1, 0);
  });

  it('setAlias calls callback', () => {
    let setAlias = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: true, isFocused: true}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        setAlias={setAlias}
        isFocused
      />
    );

    wrapper.find(EditableChip).setState({aliasMode: true});
    wrapper.find('input').simulate('keydown', {key: 'Enter', keyCode: 13});
    expect(setAlias).toBeCalledWith('ahoj', 1, 0);
  });


  it('onFocus calls callback', () => {
    let onFocus = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: false, isFocused: false}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        onFocus={onFocus}
      />
    );

    wrapper.find('span.pf-c-label').simulate('focus');
    expect(onFocus).toBeCalledWith({"id": 1, "label": "ahoj", "next": [], "parent": {"id": 0, "label": "root", "next": [], "parent": null, "type": "root"}, "type": "category"}, 1, 0);
  });

  it('OnBlur calls callback', () => {
    let onBlur = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: false, isFocused: false}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        onBlur={onBlur}
        isFocused
      />
    );

    wrapper.find('span.pf-c-label').simulate('blur');
    expect(onBlur).toBeCalledWith({"id": 1, "label": "ahoj", "next": [], "parent": {"id": 0, "label": "root", "next": [], "parent": null, "type": "root"}, "type": "category"}, 1, 0);
  });

  it('onInsertExpression calls callback', () => {
    let onInsertExpression = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: false, isFocused: false}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        onInsertExpression={onInsertExpression}
        isFocused
      />
    );

    wrapper.find('button#insert-expression-button').simulate('click');
    expect(onInsertExpression).toBeCalledWith(1);
  });

  it('onDeleteExpression calls callback', () => {
    let onDeleteExpression = jest.fn();
    const wrapper = mount(
      <Expression {...expressionProps}
        expression={[{flags: {isEditing: false, isFocused: false}, term: {id: 1, label: 'ahoj', type: 'category', next: [], parent: {id: 0, label: 'root', type: 'root', next:[], parent: null}}}]}
        onDeleteExpression={onDeleteExpression}
        isFocused
      />
    );

    wrapper.find('button#delete-expression-button').simulate('click');
    expect(onDeleteExpression).toBeCalledWith(1);
  });

  it('Matches the snapshot', () => {
    const component = shallow(
      <Expression {...expressionProps} />
    );

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  it('Matches the snapshot', () => {
    const component = shallow(
      <Expression {...expressionProps} isLastExpression parenthesesCount={{left: 1, right: 0}}/>
    );

    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
