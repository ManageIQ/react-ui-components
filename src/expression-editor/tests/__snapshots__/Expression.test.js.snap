// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expression Component Matches the snapshot 1`] = `
<div
  className="expressionRow expressionOdd"
>
  <ul
    className="list-inline expression"
  >
    <li
      className="chip"
    >
      <EditableChip
        blurAllChips={[MockFunction]}
        chipRef={
          Object {
            "current": <span
              class=""
              id="pf-toggle-id-14"
              tabindex="0"
            >
              <span
                class="pf-c-label"
              >
                ahoj
              </span>
            </span>,
          }
        }
        filterOption={[Function]}
        index={0}
        isAliasSet={false}
        isEditing={false}
        item={
          Object {
            "id": 1,
            "label": "ahoj",
            "next": Array [],
            "parent": Object {
              "id": 0,
              "label": "root",
              "next": Array [],
              "parent": null,
              "type": "root",
            },
            "type": "category",
          }
        }
        key="1"
        label="ahoj"
        onBlur={[Function]}
        onClick={[Function]}
        onDelete={[Function]}
        onDoubleClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onSubmit={[Function]}
        options={Array []}
        selected={
          Object {
            "id": 1,
            "label": "ahoj",
            "next": Array [],
            "parent": Object {
              "id": 0,
              "label": "root",
              "next": Array [],
              "parent": null,
              "type": "root",
            },
            "type": "category",
          }
        }
        setAlias={[Function]}
      />
    </li>
    <Component />
  </ul>
  <ul
    className="list-inline expressionButtons"
  >
    <li>
      <div
        className="buttons .pf-c-form-control"
      >
        <Button
          aria-label={null}
          className="button"
          component="button"
          id="delete-expression-button"
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[Function]}
          type="button"
          variant="danger"
        >
           Delete Expression 
        </Button>
        <Button
          aria-label={null}
          className="button"
          component="button"
          id="insert-expression-button"
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
           Insert Expression 
        </Button>
      </div>
    </li>
  </ul>
</div>
`;

exports[`Expression Component Matches the snapshot 2`] = `
<div
  className="expressionRow expressionOdd"
>
  <ul
    className="list-inline expression"
  >
    <li
      className="chip"
    >
      <EditableChip
        blurAllChips={[MockFunction]}
        chipRef={
          Object {
            "current": <span
              class=""
              id="pf-toggle-id-14"
              tabindex="0"
            >
              <span
                class="pf-c-label"
              >
                ahoj
              </span>
            </span>,
          }
        }
        filterOption={[Function]}
        index={0}
        isAliasSet={false}
        isEditing={false}
        item={
          Object {
            "id": 1,
            "label": "ahoj",
            "next": Array [],
            "parent": Object {
              "id": 0,
              "label": "root",
              "next": Array [],
              "parent": null,
              "type": "root",
            },
            "type": "category",
          }
        }
        key="1"
        label="ahoj"
        onBlur={[Function]}
        onClick={[Function]}
        onDelete={[Function]}
        onDoubleClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onSubmit={[Function]}
        options={Array []}
        selected={
          Object {
            "id": 1,
            "label": "ahoj",
            "next": Array [],
            "parent": Object {
              "id": 0,
              "label": "root",
              "next": Array [],
              "parent": null,
              "type": "root",
            },
            "type": "category",
          }
        }
        setAlias={[Function]}
      />
    </li>
    <Component />
  </ul>
  <ul
    className="list-inline expressionButtons"
  >
    <li>
      <div
        className="buttons .pf-c-form-control"
      >
        <Button
          aria-label={null}
          className="button"
          component="button"
          id="delete-expression-button"
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[Function]}
          type="button"
          variant="danger"
        >
           Delete Expression 
        </Button>
        <Button
          aria-label={null}
          className="button"
          component="button"
          id="insert-expression-button"
          isActive={false}
          isBlock={false}
          isDisabled={false}
          isFocus={false}
          isHover={false}
          onClick={[Function]}
          type="button"
          variant="primary"
        >
           Insert Expression 
        </Button>
      </div>
    </li>
  </ul>
</div>
`;
